File: src/ycp/enhance.py
"""Owned Ssemble-parity enhancements — pure ffmpeg, no app, no ceiling.

Replaces Ssemble's "Hook Title & CTA" and "Game Video" features with local ffmpeg.
The filter/command *builders* are pure functions (unit-tested without running
ffmpeg); the thin `apply_*` wrappers execute them.

Title/CTA text is passed to ffmpeg via `textfile=` (ffmpeg reads it from a file),
which sidesteps drawtext's fragile inline-text escaping — so titles with
apostrophes/colons (e.g. "Don't sleep on 5:00") render correctly. The font is the
vendored TikTok Sans Overlay cut (no spaces in the path), falling back to a stock
macOS font when the asset is missing.

See SSEMBLE-PARITY.md for the full map.
"""
from __future__ import annotations

import functools
import subprocess
from pathlib import Path

# Official font: TikTok Sans Overlay cut (wght 650), vendored in assets/ (path has
# no spaces — drawtext-safe). Falls back to stock macOS Helvetica if assets are absent.
_TIKTOK_OVERLAY = (Path(__file__).resolve().parents[2]
                   / "assets" / "tiktok-font" / "fonts" / "TikTokSans-Overlay.ttf")
DEFAULT_FONT = (str(_TIKTOK_OVERLAY) if _TIKTOK_OVERLAY.is_file()
                else "/System/Library/Fonts/Helvetica.ttc")


@functools.lru_cache(maxsize=None)
def ffmpeg_has_filter(name: str) -> bool:
    """True if this ffmpeg build exposes `name` (e.g. 'drawtext', 'subtitles').

    Some ffmpeg builds ship without libass/libfreetype, so drawtext/subtitles are
    absent. We detect once and let callers degrade gracefully (uncaptioned clip)
    instead of hard-failing — captions/hooks render automatically once ffmpeg has
    the text libs. See SSEMBLE-PARITY.md / setup notes for the libass install.
    """
    try:
        out = subprocess.run(["ffmpeg", "-hide_banner", "-filters"],
                             capture_output=True, text=True, timeout=30).stdout
    except (OSError, subprocess.SubprocessError):
        return False
    return any(line.split()[1:2] == [name] for line in out.splitlines() if line.strip())


def title_filter(textfile: str, font: str = DEFAULT_FONT, fontsize: int = 56) -> str:
    """Top hook-title banner (whole clip). `textfile` is a path ffmpeg reads."""
    return (f"drawtext=fontfile='{font}':textfile='{textfile}':fontcolor=white:"
            f"fontsize={fontsize}:box=1:boxcolor=black@0.55:boxborderw=18:"
            f"x=(w-text_w)/2:y=90")


def cta_filter(textfile: str, start: float, end: float, font: str = DEFAULT_FONT,
               fontsize: int = 48) -> str:
    """Timed bottom CTA banner (e.g. 'Subscribe for more'), shown start→end."""
    return (f"drawtext=fontfile='{font}':textfile='{textfile}':fontcolor=black:"
            f"fontsize={fontsize}:box=1:boxcolor=yellow@0.95:boxborderw=20:"
            f"x=(w-text_w)/2:y=h-text_h-160:enable='between(t,{start},{end})'")


def hook_cta_vf(title_file: str | None, cta_file: str | None,
                cta_window: tuple[float, float], font: str = DEFAULT_FONT) -> str:
    """Compose the title + CTA drawtext chain into one -vf string."""
    parts: list[str] = []
    if title_file:
        parts.append(title_filter(title_file, font))
    if cta_file:
        parts.append(cta_filter(cta_file, cta_window[0], cta_window[1], font))
    return ",".join(parts)


def vstack_cmd(clip: Path, gameplay: Path, out: Path,
               top_h: int = 1152, bottom_h: int = 768, width: int = 1080) -> list[str]:
    """Build the ffmpeg arg list to stack `clip` over looping `gameplay` (split-screen).

    Gameplay loops to the clip's length; clip audio is kept, gameplay audio dropped.
    """
    fc = (f"[0:v]scale={width}:{top_h}:force_original_aspect_ratio=increase,"
          f"crop={width}:{top_h}[top];"
          f"[1:v]scale={width}:{bottom_h}:force_original_aspect_ratio=increase,"
          f"crop={width}:{bottom_h},setsar=1[bot];"
          f"[top][bot]vstack=inputs=2[v]")
    return [
        "ffmpeg", "-y", "-i", str(clip), "-stream_loop", "-1", "-i", str(gameplay),
        "-filter_complex", fc, "-map", "[v]", "-map", "0:a?",
        "-c:v", "libx264", "-c:a", "aac", "-preset", "veryfast", "-shortest", str(out),
    ]


# ── thin executors ───────────────────────────────────────────────────────────

def _run(cmd: list[str], what: str, cwd: Path | None = None) -> None:
    proc = subprocess.run(cmd, capture_output=True, text=True, timeout=600, cwd=cwd)
    if proc.returncode != 0:
        raise RuntimeError(f"{what} failed: {proc.stderr.strip()[-400:]}")


def apply_overlay(video: Path, out: Path, title: str | None = None, cta: str | None = None,
                  cta_window: tuple[float, float] = (2.0, 7.0), font: str = DEFAULT_FONT) -> Path:
    """Burn a hook title + CTA banner. Text is written to files next to `out` and
    referenced via textfile= (escaping-proof). Runs ffmpeg with cwd=out.parent."""
    if not ffmpeg_has_filter("drawtext"):
        print("  ⚠ ffmpeg lacks drawtext (no libfreetype) — skipping hook/CTA overlay; "
              "reinstall ffmpeg with libfreetype/libass to burn titles")
        return video
    workdir = out.parent
    title_file = cta_file = None
    if title:
        (workdir / "title.txt").write_text(title)
        title_file = "title.txt"
    if cta:
        (workdir / "cta.txt").write_text(cta)
        cta_file = "cta.txt"
    vf = hook_cta_vf(title_file, cta_file, cta_window, font)
    if not vf:
        return video
    _run(["ffmpeg", "-y", "-i", str(video), "-vf", vf, "-c:v", "libx264",
          "-c:a", "copy", "-preset", "veryfast", str(out)], "overlay", cwd=workdir)
    return out


def stack_gameplay(clip: Path, gameplay: Path, out: Path) -> Path:
    if not gameplay.exists():
        raise FileNotFoundError(f"gameplay loop not found: {gameplay}")
    _run(vstack_cmd(clip, gameplay, out), "gameplay vstack")
    return out


def pick_title(transcript: str, max_words: int = 9) -> str:
    """Heuristic hook title from the transcript: first question, else punchiest line.

    Zero-dependency fallback used when the DeepSeek hook agent (hooks.best_hook)
    is unavailable — e.g. no DEEPSEEK_API_KEY configured for the run.
    """
    sentences = [s.strip() for s in transcript.replace("!", ".").replace("?", "?.").split(".")
                 if s.strip()]
    if not sentences:
        return ""
    question = next((s for s in sentences if s.endswith("?")), None)
    pick = question or max(sentences, key=len)
    words = pick.split()
    return " ".join(words[:max_words]) + ("…" if len(words) > max_words else "")

File: rust/src/enhance.rs
//! Owned ffmpeg enhancements — parity port of `src/ycp/enhance.py`.
//!
//! `pick_title` is the zero-dependency hook heuristic `hooks::best` falls back to.
//! `stack_gameplay` (split-screen retention) shells out to ffmpeg, mirroring `vstack_cmd`.

use std::path::{Path, PathBuf};
use std::process::Command;

use anyhow::{bail, Result};

/// ffmpeg args to stack `clip` over looping `gameplay` (split-screen). Mirrors `vstack_cmd`:
/// gameplay loops to the clip's length, clip audio kept, gameplay audio dropped. Pure builder.
pub fn vstack_cmd(clip: &Path, gameplay: &Path, out: &Path) -> Vec<String> {
    let (top_h, bottom_h, width) = (1152, 768, 1080);
    let fc = format!(
        "[0:v]scale={width}:{top_h}:force_original_aspect_ratio=increase,crop={width}:{top_h}[top];\
         [1:v]scale={width}:{bottom_h}:force_original_aspect_ratio=increase,crop={width}:{bottom_h},setsar=1[bot];\
         [top][bot]vstack=inputs=2[v]"
    );
    vec![
        "-y".into(),
        "-i".into(),
        clip.display().to_string(),
        "-stream_loop".into(),
        "-1".into(),
        "-i".into(),
        gameplay.display().to_string(),
        "-filter_complex".into(),
        fc,
        "-map".into(),
        "[v]".into(),
        "-map".into(),
        "0:a?".into(),
        "-c:v".into(),
        "libx264".into(),
        "-c:a".into(),
        "aac".into(),
        "-preset".into(),
        "veryfast".into(),
        "-shortest".into(),
        out.display().to_string(),
    ]
}

/// Stack `clip` over a looping `gameplay` file (split-screen retention). Mirrors `stack_gameplay`.
pub fn stack_gameplay(clip: &Path, gameplay: &Path, out: &Path) -> Result<PathBuf> {
    if !gameplay.exists() {
        bail!("gameplay loop not found: {}", gameplay.display());
    }
    let res = Command::new("ffmpeg")
        .args(vstack_cmd(clip, gameplay, out))
        .output()?;
    if !res.status.success() {
        let err = String::from_utf8_lossy(&res.stderr);
        bail!(
            "gameplay vstack failed: {}",
            err[err.len().saturating_sub(400)..].trim()
        );
    }
    Ok(out.to_path_buf())
}

/// Heuristic hook title from the transcript: first question, else the longest line.
///
/// Mirrors enhance.py `pick_title` — zero-dependency fallback when the DeepSeek hook
/// agent is unavailable (no DEEPSEEK_API_KEY).
pub fn pick_title(transcript: &str, max_words: usize) -> String {
    // Python: transcript.replace("!", ".").replace("?", "?.").split(".") then strip+drop-empty.
    let replaced = transcript.replace('!', ".").replace('?', "?.");
    let sentences: Vec<&str> = replaced
        .split('.')
        .map(str::trim)
        .filter(|s| !s.is_empty())
        .collect();
    if sentences.is_empty() {
        return String::new();
    }
    // First question wins; else max(by char-length). Python `max` returns the FIRST max on
    // ties, so keep only strictly-longer to match.
    let pick = sentences
        .iter()
        .find(|s| s.ends_with('?'))
        .copied()
        .unwrap_or_else(|| {
            let mut best = sentences[0];
            for &s in &sentences[1..] {
                if s.chars().count() > best.chars().count() {
                    best = s;
                }
            }
            best
        });
    let words: Vec<&str> = pick.split_whitespace().collect();
    let joined = words
        .iter()
        .take(max_words)
        .copied()
        .collect::<Vec<_>>()
        .join(" ");
    if words.len() > max_words {
        format!("{joined}…")
    } else {
        joined
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn prefers_the_question() {
        let out = pick_title("Why does nobody talk about this? It changes everything.", 9);
        assert_eq!(out, "Why does nobody talk about this?");
    }

    #[test]
    fn falls_back_to_longest_line() {
        let out = pick_title("Short. The longest sentence here is this one. Mid line.", 9);
        assert_eq!(out, "The longest sentence here is this one");
    }

    #[test]
    fn empty_transcript_is_empty() {
        assert_eq!(pick_title("   ", 9), "");
    }

    #[test]
    fn trims_to_max_words_with_ellipsis() {
        let out = pick_title("one two three four five six seven eight", 3);
        assert_eq!(out, "one two three…");
    }
}
